File: VB/Q245442/Collections/SortableBindingList.vb
Imports System.ComponentModel
Imports System.Collections
Imports DevExpress.Data
Imports System.Windows.Forms
Imports System
Imports DevExpress.Data.Helpers

Namespace Q245442.Collections
    Public Class SortableBindingList
        Implements IBindingList

        Private DataController As ListSourceDataController

        Public Sub New(ByVal bindingContext As BindingContext, ByVal sourceList As IList)
            Me.DataController = New ListSourceDataController()
            Me.DataController.SetListSource(bindingContext, sourceList, String.Empty)
        End Sub

        Protected Overridable Sub ApplySort(ByVal columnInfo As DataColumnInfo, ByVal direction As ListSortDirection)
            Dim sortOrder As ColumnSortOrder = ColumnSortOrder.Ascending
            If direction = ListSortDirection.Descending Then
                sortOrder = ColumnSortOrder.Descending
            End If
            Dim sortInfo As New DataColumnSortInfo(columnInfo, sortOrder)
            Me.DataController.SortInfo.ClearAndAddRange(sortInfo)
        End Sub

        Public Overridable Sub RemoveSort()
            Me.DataController.SortInfo.Clear()
        End Sub

        Public Sub ApplySort(ByVal [property] As String, ByVal direction As ListSortDirection)
            Dim column As DataColumnInfo = Me.DataController.FindColumn([property])
            If column Is Nothing Then
                Return
            End If
            Me.ApplySort(column, direction)
        End Sub

        #Region "IBindingList members"
        Private Sub IBindingList_AddIndex(ByVal [property] As PropertyDescriptor) Implements IBindingList.AddIndex
        End Sub

        Private Function IBindingList_AddNew() As Object Implements IBindingList.AddNew
            Throw New NotSupportedException("Adding a new item to the list is not supported")
        End Function

        Private ReadOnly Property IBindingList_AllowEdit() As Boolean Implements IBindingList.AllowEdit
            Get
                Return True
            End Get
        End Property

        Private ReadOnly Property IBindingList_AllowNew() As Boolean Implements IBindingList.AllowNew
            Get
                Return False
            End Get
        End Property

        Private ReadOnly Property IBindingList_AllowRemove() As Boolean Implements IBindingList.AllowRemove
            Get
                Return False
            End Get
        End Property

        Private Sub IBindingList_ApplySort(ByVal [property] As PropertyDescriptor, ByVal direction As ListSortDirection) Implements IBindingList.ApplySort
            Dim columnInfo As DataColumnInfo = Me.DataController.FindColumn([property].Name)
            If columnInfo Is Nothing Then
                Return
            End If
            Me.ApplySort(columnInfo, direction)
        End Sub

        Private Function IBindingList_Find(ByVal [property] As PropertyDescriptor, ByVal key As Object) As Integer Implements IBindingList.Find
            Throw New NotSupportedException("Searching is not supported")
        End Function

        Private ReadOnly Property IBindingList_IsSorted() As Boolean Implements IBindingList.IsSorted
            Get
                Return Me.DataController.IsSorted
            End Get
        End Property

        Private Custom Event ListChanged As ListChangedEventHandler Implements IBindingList.ListChanged
            AddHandler(ByVal value As ListChangedEventHandler)
            End AddHandler
            RemoveHandler(ByVal value As ListChangedEventHandler)
            End RemoveHandler
            RaiseEvent(ByVal sender As System.Object, ByVal e As System.ComponentModel.ListChangedEventArgs)
            End RaiseEvent
        End Event

        Private Sub IBindingList_RemoveIndex(ByVal [property] As PropertyDescriptor) Implements IBindingList.RemoveIndex
        End Sub

        Private Sub IBindingList_RemoveSort() Implements IBindingList.RemoveSort
            Me.RemoveSort()
        End Sub

        Private ReadOnly Property IBindingList_SortDirection() As ListSortDirection Implements IBindingList.SortDirection
            Get
                Dim result As ListSortDirection = ListSortDirection.Ascending
                If Me.DataController.SortInfo(0).SortOrder = ColumnSortOrder.Descending Then
                    result = ListSortDirection.Descending
                End If
                Return result
            End Get
        End Property

        Private ReadOnly Property IBindingList_SortProperty() As PropertyDescriptor Implements IBindingList.SortProperty
            Get
                Return Me.DataController.SortInfo(0).ColumnInfo.PropertyDescriptor
            End Get
        End Property

        Private ReadOnly Property IBindingList_SupportsChangeNotification() As Boolean Implements IBindingList.SupportsChangeNotification
            Get
                Return False
            End Get
        End Property

        Private ReadOnly Property IBindingList_SupportsSearching() As Boolean Implements IBindingList.SupportsSearching
            Get
                Return False
            End Get
        End Property

        Private ReadOnly Property IBindingList_SupportsSorting() As Boolean Implements IBindingList.SupportsSorting
            Get
                Return True
            End Get
        End Property
        #End Region
        #Region "IList members"
        Private Function IList_Add(ByVal value As Object) As Integer Implements IList.Add
            Throw New NotSupportedException("The list is read-only")
        End Function

        Private Sub IList_Clear() Implements IList.Clear
            Throw New NotSupportedException("The list is read-only")
        End Sub

        Private Function IList_Contains(ByVal value As Object) As Boolean Implements IList.Contains
            Return Me.DataController.ListSource.Contains(value)
        End Function

        Private Function IList_IndexOf(ByVal value As Object) As Integer Implements IList.IndexOf
            Return Me.DataController.ListSource.IndexOf(value)
        End Function

        Private Sub IList_Insert(ByVal index As Integer, ByVal value As Object) Implements IList.Insert
            Throw New NotSupportedException("The list is read-only")
        End Sub

        Private ReadOnly Property IList_IsFixedSize() As Boolean Implements IList.IsFixedSize
            Get
                Return Me.DataController.ListSource.IsFixedSize
            End Get
        End Property

        Private ReadOnly Property IList_IsReadOnly() As Boolean Implements IList.IsReadOnly
            Get
                Return True
            End Get
        End Property

        Private Sub IList_Remove(ByVal value As Object) Implements IList.Remove
            Throw New NotSupportedException("The list is read-only")
        End Sub

        Private Sub IList_RemoveAt(ByVal index As Integer) Implements IList.RemoveAt
            Throw New NotSupportedException("The list is read-only")
        End Sub

        Public Property IList_Item(ByVal index As Integer) As Object Implements IList.Item
            Get
                Return Me.DataController.GetRow(index)
            End Get
            Set(ByVal value As Object)
                Throw New NotSupportedException("The list is read-only")
            End Set
        End Property
        #End Region
        #Region "ICollection members"
        Private Sub ICollection_CopyTo(ByVal array As Array, ByVal index As Integer) Implements ICollection.CopyTo
            Me.DataController.ListSource.CopyTo(array, index)
        End Sub

        Private ReadOnly Property ICollection_Count() As Integer Implements ICollection.Count
            Get
                Return Me.DataController.VisibleCount
            End Get
        End Property

        Private ReadOnly Property ICollection_IsSynchronized() As Boolean Implements ICollection.IsSynchronized
            Get
                Return Me.DataController.ListSource.IsSynchronized
            End Get
        End Property

        Private ReadOnly Property ICollection_SyncRoot() As Object Implements ICollection.SyncRoot
            Get
                Return Me.DataController.ListSource.SyncRoot
            End Get
        End Property
        #End Region
        #Region "IEnumerator members"
        Private Function IEnumerable_GetEnumerator() As IEnumerator Implements IEnumerable.GetEnumerator
            Return Me.DataController.ListSource.GetEnumerator()
        End Function
        #End Region
    End Class
End Namespace

File: CS/Q245442/Collections/SortableBindingList.cs
using System.ComponentModel;
using System.Collections;
using DevExpress.Data;
using System.Windows.Forms;
using System;
using DevExpress.Data.Helpers;

namespace Q245442.Collections {
    public class SortableBindingList : IBindingList {
        private ListSourceDataController DataController;

        public SortableBindingList(BindingContext bindingContext, IList sourceList) {
            this.DataController = new ListSourceDataController();
            this.DataController.SetListSource(bindingContext, sourceList, string.Empty);
        }

        protected virtual void ApplySort(DataColumnInfo columnInfo, ListSortDirection direction) {
            ColumnSortOrder sortOrder = ColumnSortOrder.Ascending;
            if (direction == ListSortDirection.Descending)
                sortOrder = ColumnSortOrder.Descending;
            DataColumnSortInfo sortInfo = new DataColumnSortInfo(columnInfo, sortOrder);
            this.DataController.SortInfo.ClearAndAddRange(sortInfo);
        }

        public virtual void RemoveSort() {
            this.DataController.SortInfo.Clear();
        }

        public void ApplySort(string property, ListSortDirection direction) {
            DataColumnInfo column = this.DataController.FindColumn(property);
            if (column == null) return;
            this.ApplySort(column, direction); 
        }

        #region IBindingList members
        void IBindingList.AddIndex(PropertyDescriptor property) { }

        object IBindingList.AddNew() {
            throw new NotSupportedException("Adding a new item to the list is not supported");
        }

        bool IBindingList.AllowEdit {
            get { return true; }
        }

        bool IBindingList.AllowNew {
            get { return false; }
        }

        bool IBindingList.AllowRemove {
            get { return false; }
        }

        void IBindingList.ApplySort(PropertyDescriptor property, ListSortDirection direction) {
            DataColumnInfo columnInfo = this.DataController.FindColumn(property.Name);
            if (columnInfo == null) return;
            this.ApplySort(columnInfo, direction);
        }

        int IBindingList.Find(PropertyDescriptor property, object key) {
            throw new NotSupportedException("Searching is not supported");
        }

        bool IBindingList.IsSorted {
            get { return this.DataController.IsSorted; }
        }

        event ListChangedEventHandler IBindingList.ListChanged {
            add { }
            remove { }
        }

        void IBindingList.RemoveIndex(PropertyDescriptor property) { }

        void IBindingList.RemoveSort() {
            this.RemoveSort();
        }

        ListSortDirection IBindingList.SortDirection {
            get {
                ListSortDirection result = ListSortDirection.Ascending;
                if (this.DataController.SortInfo[0].SortOrder == ColumnSortOrder.Descending)
                    result = ListSortDirection.Descending;
                return result;
            }
        }

        PropertyDescriptor IBindingList.SortProperty {
            get { return this.DataController.SortInfo[0].ColumnInfo.PropertyDescriptor; }
        }

        bool IBindingList.SupportsChangeNotification {
            get { return false; }
        }

        bool IBindingList.SupportsSearching {
            get { return false; }
        }

        bool IBindingList.SupportsSorting {
            get { return true; }
        }
        #endregion
        #region IList members
        int IList.Add(object value) {
            throw new NotSupportedException("The list is read-only");
        }

        void IList.Clear() {
            throw new NotSupportedException("The list is read-only");
        }

        bool IList.Contains(object value) {
            return this.DataController.ListSource.Contains(value);
        }

        int IList.IndexOf(object value) {
            return this.DataController.ListSource.IndexOf(value);
        }

        void IList.Insert(int index, object value) {
            throw new NotSupportedException("The list is read-only");
        }

        bool IList.IsFixedSize {
            get { return this.DataController.ListSource.IsFixedSize; }
        }

        bool IList.IsReadOnly {
            get { return true; }
        }

        void IList.Remove(object value) {
            throw new NotSupportedException("The list is read-only");
        }

        void IList.RemoveAt(int index) {
            throw new NotSupportedException("The list is read-only");
        }

        object IList.this[int index] {
            get { return this.DataController.GetRow(index); }
            set { throw new NotSupportedException("The list is read-only"); }
        }
        #endregion
        #region ICollection members
        void ICollection.CopyTo(Array array, int index) {
            this.DataController.ListSource.CopyTo(array, index);
        }

        int ICollection.Count {
            get { return this.DataController.VisibleCount; }
        }

        bool ICollection.IsSynchronized {
            get { return this.DataController.ListSource.IsSynchronized; }
        }

        object ICollection.SyncRoot {
            get { return this.DataController.ListSource.SyncRoot; }
        }
        #endregion
        #region IEnumerator members
        IEnumerator IEnumerable.GetEnumerator() {
            return this.DataController.ListSource.GetEnumerator();
        }
        #endregion
    }
}
